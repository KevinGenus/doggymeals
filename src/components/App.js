import React, { Component } from "react";
import { Button } from "react-bootstrap";
import Meal from "./Meal";
import { max_number } from "../helper";

class App extends Component {
  constructor() {
    super();
    this.state = { meals: [] };
  }

  addMeal = () => {
    const { meals } = this.state;
    meals.push({ id: max_number(this.state.meals.map(meal => meal.id)) + 1 });
    this.setState({ meals });
  };

  removeMeal = id => {
    const meals = this.state.meals.filter(meal => meal.id !== id);
    this.setState({ meals });
  };

  render() {
    return (
      <div>
        <div>Doggy Meals</div>
        <div className="meal-list">
          {this.state.meals.map(meal => {
            return (
              <Meal key={meal.id} meal={meal} removeMeal={this.removeMeal} />
            );
          })}
        </div>
        <Button onClick={this.addMeal} className="btn-add">
          Add Doggy Meal
        </Button>
      </div>
    );
  }
}

export default App;
