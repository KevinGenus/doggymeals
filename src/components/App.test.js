import React from "react";
import { shallow } from "enzyme";
import App from "./App";
import toJson from "enzyme-to-json";

describe("App", () => {
  const app = shallow(<App />);

  it("renders correctly", () => {
    expect(toJson(app)).toMatchSnapshot();
  });

  it("initializes the `state` with an empty list of meals", () => {
    expect(app.state().meals).toEqual([]);
  });

  describe("when clicking the `add-meal` button", () => {
    const id = 1;
    beforeEach(() => {
      app.find(".btn-add").simulate("click");
    });

    afterEach(() => {
      app.setState({ meals: [] });
    });

    it("adds a new gift to `state` when clicking the `add meal` button", () => {
      expect(app.state().meals).toEqual([{ id }]);
    });

    it("adds a new meal to the rendered list when clicked the `add meal` button", () => {
      expect(app.find(".meal-list").children().length).toEqual(1);
    });

    it("creates a Meal component", () => {
      expect(app.find("Meal").exists()).toBe(true);
    });

    describe("and the user wants to remove the added meal", () => {
      beforeEach(() => {
        app.instance().removeMeal(id);
      });

      it("removes the meal from `state`", () => {
        expect(app.state().meals).toEqual([]);
      });
    });
  });
});
