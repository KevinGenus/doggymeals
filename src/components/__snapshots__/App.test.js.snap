// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div>
  <div>
    Doggy Meals
  </div>
  <div
    className="meal-list"
  />
  <Button
    active={false}
    className="btn-add"
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Add Doggy Meal
  </Button>
</div>
`;
